api/src/main/java/bio/terra/tanagra/serialization/UFAttribute.java

```diff
@@ -0,0 +1,85 @@
+package bio.terra.tanagra.serialization;
```

Comment:
I think Java records looks promising. I'm not sure if it allows inheritance (e.g. each data pointer class extends a base one), though we don't use that too much in the serialization classes so maybe not a dealbreaker.

However, records are available in Java 14+ and Tanagra is currently at Java 11. I want to tackle a version upgrade in a separate PR, since this overhaul [part 1] is already very big. That's why I went with POJOs here.

I think decoupling the user-facing config schema from the internal service/indexing logic is useful. We can be more flexible to requested changes to the config schema, and our internal logic classes are insulated from that.

There is a lot of conversion code, but it's not just applying default values. It's also passing references throughout, to reduce repetition in the user-defined config (e.g. for an attribute that maps to a simple column, the query engine needs the BQ project+dataset id+table name+column name, but we only want the user to specify the column name in the config, and get the rest of the information from higher up). I think we'd need that part regardless.

Reply:
I guess leave it for now but this is the downside to massive PRs like this. If records are the way to go it would make more sense to do the version upgrade first rather than do this poorly and go back to it but I understand that may be painful to do with all of this code to rebase on top.

Reply:
Yeah I actually tried to do the version upgrade first, but it gave me a bunch of errors in the Apache Beam jobs. And given the way the code is structured (1 Gradle project, with the api and workflows in separate sub-projects), we can't upgrade in just the api and leave the workflow unchanged. So I decided to just leave it and focus on the underlay overhaul. I still think we should upgrade, but priority-wise it should come after all our feature work.